### ISSUE TITLE
Non-preempting PriorityClass

### ISSUE THREAD

**Issue by bsalamat** [kubernetes/kubernetes#67671]
As of Kubernetes 1.11, [Pod priority and preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) is a beta feature. In the existing version, any Pod could possibly preempt other Pods if it has higher priority than the preemption victims and the cluster does not have enough capacity to run it.
We would like to add a new feature that allows certain PriorityClasses to be defined as non-preempting priority. When a Pod has non-preempting priority, it won't cause preemption when the cluster does not have capacity to run it, but it is still ordered by its priority in the scheduling queue. So, it has a higher chance of getting scheduled than other pending pods with lower priority.

One common use-case for such non-preempting priorities is batch processing. Typical batch jobs run to completion (as opposed to services that keep running). In batch processing, if preemption happens, all the work done by the preemption victim will be lost. As a result, it is often not desired to preempt the existing workload, but priority is still needed to ensure that when new workloads are created those with higher priority go in front the queue and are scheduled ahead of other lower priority workloads when the cluster has more free resources, for example, after currently running workloads are finished.

**Implementation:**
We can add a new field, called `non-preempting` to [PriorityClass](https://github.com/kubernetes/kubernetes/blob/7749eeaf3f1eddd0796a4d61e545a90e7d9d59b0/staging/src/k8s.io/api/scheduling/v1beta1/types.go#L29). When a pod is unschedulable, the scheduler checks this field of its corresponding PriorityClass and starts the preemption process when the field is empty or false. Several points to be considered:
1. All PriorityClasses are considered `preempting` by default. If the filed is missing, the PriorityClass is a preempting one.
2.  A PriorityClass may be deleted after Pod creation. If scheduler cannot find the PriorityClass of a Pod, it assumes that it has `preempting` priority.
3. A Pod may not have any PriorityClass. In that case, its priority is resolved to a defined or implicit default priority. When a Pod priority class is empty, its priority is considered as `preempting`.

To summarize the above rules: "A Pod has `preempting` priority, unless its PriorityClass is found and the `non-preempting` field of the PriorityClass is set to true".

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

/ref #64843
/sig scheduling

**Comment by tanshanshan** [comment]
I want to finish it . could you assign to me ?

**Comment by liggitt** [comment]
> the scheduler checks this field of its corresponding PriorityClass

the scheduler doesn't currently look at priority class, only at pods. splitting the data between pods and priority class would mean looking at the pod would be insufficient for the scheduler. why would the same approach of denormalizing priority value to the pod spec on creation not be followed?

**Comment by bsalamat** [comment]
Thanks, @tanshanshan! Assigned.

@liggitt That would need another field in the PodSpec. This feature may not be the last feature added to PriorityClass. We cannot keep adding new fields to PodSpec. Scheduler looks at many other objects than just Pods anyway. PodDisruptionBudget, Node, PV, PVC are examples of objects that scheduler monitors.

**Comment by k82cn** [comment]
> This feature may not be the last feature added to PriorityClass. We cannot keep adding new fields to PodSpec. 

Should we watch PriorityClass in scheduler, and handle it in scheduler.cache?

**Comment by bsalamat** [comment]
> Should we watch PriorityClass in scheduler, and handle it in scheduler.cache?

I think we don't need to keep PriorityClasses in the scheduler's cache. We can rely on the informer cache for PriorityClasses.

**Comment by k82cn** [comment]
oh, I means we introduce `PodInfo` in scheduler.cache to include PriorityClass and Pod :) So we do not need to update PodSpec when introducing new field in PriorityClass.

Can not remember whether we discuss this in design doc: if client/user create several Jobs/RSs,  and low priority first, the high priority job/rs may not create pod because of admission controller so the scheduler can not do preemption.

**Comment by nikhita** [comment]
@tanshanshan Are you working or do you plan on working on this? If yes, could you please remove the help wanted label? Thanks!

**Comment by bsalamat** [comment]
@vllry there is a [PR](https://github.com/kubernetes/kubernetes/pull/67742) which is incomplete as it causes build failures. I never found the time to investigate further what is causing the build failure, but feel free to take a look and fix it. At this point, I don't think I will find the time to provide the solution.

**Comment by tanshanshan** [comment]
@bsalamat  @k82cn 
sorry , but I feel so difficult to solve it ,so I closed it .
sorry for delaying long time.

**Comment by liggitt** [comment]
@bsalamat is this a fundamental use case we need to address before [promoting the API to v1](https://github.com/kubernetes/kubernetes/pull/73555)?

**Comment by bsalamat** [comment]
@liggitt This is not a fundamental use case, but it is a good enhancement. This does not need to be addressed before promoting the API to V1.

**Comment by denkensk** [comment]
@vllry  Are you still working on this? If not, maybe I can have a try? Thanks!

**Comment by vllry** [comment]
@denkensk I just finished moving so I haven’t had much time... I’m trying to get back into it but you’re free to take over if you want. I’m struggling to track down why the TestRoundTrip is failing.

**Comment by denkensk** [comment]
@vllry  Sure, thank you. Let me take over it and have a try~ 
/assign @denkensk 
/unassign @tanshanshan

**Comment by vllry** [comment]
I'd like to suggest switching the name of the field to Preempting (default true), rather than NonPreempting (default false). It's more clear to users.

The API docs indicate that we can tell the difference between unset and false values, so we don't need the default boolean to be false. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md

**Comment by bsalamat** [comment]
> I'd like to suggest switching the name of the field to Preempting (default true), rather than NonPreempting (default false). It's more clear to users.

I agree. Thanks for the suggestion.

**Comment by denkensk** [comment]
> I'd like to suggest switching the name of the field to Preempting (default true), rather than NonPreempting (default false). It's more clear to users.

Thanks for the suggestion. I will update the code soon.
